src/mmap.cpp

```diff
@@ -117,6 +177,29 @@ file_handle::file_handle(string_view name, std::int64_t
 		DWORD temp;
 		::DeviceIoControl(m_fd, FSCTL_SET_SPARSE, nullptr, 0, nullptr, 0, &temp, nullptr);
 	}
+
+	if (mode & open_mode::truncate)
+	{
+		LARGE_INTEGER sz;
+		sz.QuadPart = size;
+		if (SetFilePointerEx(m_fd, sz, nullptr, FILE_BEGIN) == FALSE)
+			throw_ex<system_error>(error_code(GetLastError(), system_category()));
+
+		if (::SetEndOfFile(m_fd) == FALSE)
+			throw_ex<system_error>(error_code(GetLastError(), system_category()));
+
+		if (!(mode & aux::open_mode::sparse))
+		{
+			// Enable privilege required by SetFileValidData()
+			// https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-setfilevaliddata
+			std::call_once(g_once_flag, acquire_manage_volume_privs);
+
+			// if the user has permissions, avoid filling
+			// the file with zeroes, but just fill it with
+			// garbage instead
+			SetFileValidData(m_fd, size);
```

Comment:
@FranciscoPombal @ssiloti
do you think it's worth preserving this feature? I believe it only works if you run as admin, in which case it can significantly speed up allocating storage on disk, since it will skip zeroing out data (and instead leave old garbage in the new file, leading to a security risk).

Reply:
@arvidn If this is about overwriting data of a deleted file:

I think the expectation of "_securely_ erasing files/data" is out-of-scope for libtorrent, so it is reasonable to keep this as a potential performance optimization.

Reply:
well, it's about how someone running a libtorrent process as admin may bring files into existence that have previously-deleted (and possibly sensitive) data already in them. To users not familiar with this, it can be quite surprising. But then again, these days, having a user run as admin would *also* be quite surprising.

Early versions of the Steam client used to create its cache files like this (back when everyone was admin on Win XP) and there was quite a stir in the forums with users thinking that the client was spying on them and storing unrelated personal data in its cache file.

Reply:
@arvidn ah, I see.

> But then again, these days, having a user run as admin would also be quite surprising.

:+1:

If someone is running software that uses libtorrent as admin, they either trust it, and realize this isn't a problem, or they don't, but all bets are off anyway since they are running as admin.